### PR TITLE
test: add integration tests for access control and negative flows

### LINKED ISSUE #132: chore: add Demo 1 backend integration test coverage
## Summary

Add database-backed backend integration test coverage for the Demo 1 flows.

The backend already has unit/API-style tests that mock or isolate parts of the stack. This chore should extend the existing integration-test skeleton so Demo 1 routes are verified against the real Prisma schema, committed migrations, repositories, services, Express routes, and relational PostgreSQL data.

**Suggested branch name:** `chore/backend-integration-tests/rudolph`

**Story Points:** 5

## Reason

Mocked route/unit tests are useful, but they do not prove that the implemented backend works against the real database model.

Database-backed integration tests should give confidence that:

- Prisma queries match `apps/backend/prisma/schema.prisma`.
- Committed migrations can be applied to the test database.
- Demo 1 relational data can be created with the actual schema.
- Route, middleware, service, repository, Prisma, and database behavior works together.
- Access-control checks work with real trainee, assignment, campaign item, and component relationships.
- Enum values and nullable fields match the database.
- Quiz attempts, answers, answer options, and results persist correctly.
- Training viewed/completed interactions persist correctly.
- Simulated inbox, email detail, interaction, and classification flows work from relational data.

The repo already has a starting point for this work:

- `apps/backend/vitest.integration.config.ts`
- `apps/backend/tests/setup.integration.ts`
- `apps/backend/tests/helpers/database.ts`
- `apps/backend/tests/integration/test-database.test.ts`
- `apps/backend/TESTING.md`
- backend script: `test:integration`

This chore should build on that setup rather than replacing existing unit/API tests.

## Scope

Included:

- Confirm and reuse the existing backend integration test setup.
- Keep integration tests on a real isolated PostgreSQL test database via `TEST_DATABASE_URL`.
- Ensure the test database uses committed Prisma migrations before tests run.
- Add reusable test data helpers/factories for Demo 1 relational entities.
- Add integration tests for auth:
  - `POST /auth/register`
  - `POST /auth/login`
  - `GET /auth/me`
- Add UC-01 simulated inbox/email integration tests:
  - `GET /trainee/campaign-items/:campaignItemId/simulated-inbox`
  - `GET /trainee/campaign-items/:campaignItemId/simulated-emails/:emailId`
  - `POST /trainee/campaign-items/:campaignItemId/simulated-emails/:emailId/interactions`
  - `POST /trainee/campaign-items/:campaignItemId/simulated-emails/:emailId/classification`
- Add UC-02 training document integration tests:
  - `GET /trainee/campaign-items/:campaignItemId/training-document`
  - `POST /trainee/campaign-items/:campaignItemId/training-document/viewed`
  - `POST /trainee/campaign-items/:campaignItemId/training-document/completed`
- Add UC-03 quiz integration tests:
  - `GET /trainee/campaign-items/:campaignItemId/quiz`
  - `POST /trainee/campaign-items/:campaignItemId/quiz/attempts`
  - `POST /quiz-attempts/:attemptId/submit`
  - `GET /quiz-attempts/:attemptId/results`
- Add access-control tests for a trainee trying to access another trainee’s campaign item, quiz attempt, training document, simulated inbox, or simulated email data.
- Add at least one database-backed not-found or forbidden path per major Demo 1 flow.
- Verify HTTP responses and persisted database state where relevant.
- Ensure integration tests clean or isolate state between tests using the existing test database helper.
- Update backend testing documentation only if new setup steps or environment variables are introduced.
- Update CI so backend integration tests run against a PostgreSQL test database.
- Ensure CI clearly separates backend unit test execution from backend integration test execution.

Out of scope:

- Replacing existing unit/API tests.
- Exhaustive DTO validation coverage already handled by unit/API tests.
- Full frontend integration or e2e testing.
- Playwright/browser testing.
- Performance/load testing.
- Large seed-data redesign unless required for readable integration factories.
- Frontend CI changes unrelated to backend integration coverage.

## CI Updates

Update CI so backend unit tests and backend integration tests are separate, clearly named checks.

This matters because a unit test failure and an integration test failure usually point to different problems. Unit failures usually indicate isolated route, service, mapper, or validation regressions. Integration failures may indicate database setup, migrations, Prisma schema mismatches, relational data issues, or end-to-end backend flow regressions.

Expected CI changes:

- Keep backend unit tests in their own CI step/check, for example:
  - `Backend unit tests`
  - command: `pnpm --filter @insightful-phish/backend test:unit`
- Add backend integration tests as a separate CI step/check, for example:
  - `Backend integration tests`
  - command: `pnpm --filter @insightful-phish/backend test:integration`
- Do not hide integration tests inside the same unnamed command as unit tests.
- Add or reuse a PostgreSQL service for the backend integration test step.
- Create or use a dedicated CI test database whose name contains `test`, for example `insightful_phish_test`.
- Set `TEST_DATABASE_URL` for the backend integration test step.
- Run Prisma generation before integration tests:
  - `pnpm --filter @insightful-phish/backend prisma:generate`
- Apply committed migrations to the CI test database before integration tests:
  - `pnpm --filter @insightful-phish/backend exec prisma migrate deploy`
- Run backend integration tests only after the PostgreSQL service is reachable and migrations have been applied.
- Make CI output/check names clear enough that reviewers can immediately tell whether a failure is from:
  - backend unit tests
  - backend integration test setup
  - backend integration test assertions

The CI database URL should point only to a disposable test database. It must satisfy the existing cleanup guard in `apps/backend/tests/helpers/database.ts`, which rejects non-test database names and protected databases such as `insightful_phish_dev`.

## Files / Areas to Touch

- `.github/workflows/ci.yml`
- `apps/backend/vitest.integration.config.ts`
- `apps/backend/tests/setup.integration.ts`
- `apps/backend/tests/helpers/database.ts`
- `apps/backend/tests/integration/test-database.test.ts`
- `apps/backend/tests/integration/**/*.test.ts`
- `apps/backend/tests/helpers/**/*`
- `apps/backend/TESTING.md`
- `apps/backend/package.json`, only if scripts need adjustment
- `apps/backend/prisma/schema.prisma`, only if tests expose a real schema issue
- `apps/backend/prisma/migrations/**/*`, only if tests expose a real migration issue

## Risks

- Integration tests depend on a local or CI PostgreSQL test database being available.
- Cleanup must never run against `insightful_phish_dev` or any non-test database.
- CI must use a disposable database name that passes the cleanup guard.
- Shared database state can make tests flaky if cleanup/factories are incomplete.
- Rate limiting middleware may affect repeated Supertest calls unless tests use isolated users or account for limits.
- Tests that depend too heavily on seed data may become brittle; prefer explicit factories per test.
- CI runtime may increase because migrations and database-backed tests are slower than unit tests.
- If CI combines unit and integration tests into one check, failures will be harder to triage.

## Tasks

- [ ] Inspect the existing backend integration test skeleton.
- [ ] Confirm `TEST_DATABASE_URL` and cleanup behavior are documented and safe.
- [ ] Confirm integration tests use the real Prisma client and not mocked Prisma.
- [ ] Confirm committed migrations apply cleanly to the test database.
- [ ] Add reusable Demo 1 test data helpers/factories.
- [ ] Add auth integration tests for register, login, and me.
- [ ] Add UC-01 integration tests for simulated inbox retrieval.
- [ ] Add UC-01 integration tests for simulated email detail retrieval.
- [ ] Add UC-01 integration tests for interaction persistence.
- [ ] Add UC-01 integration tests for classification persistence and duplicate classification behavior.
- [ ] Add UC-02 integration tests for fetching a training document.
- [ ] Add UC-02 integration tests for marking a training document as viewed.
- [ ] Add UC-02 integration tests for marking a training document as completed.
- [ ] Add UC-03 integration tests for fetching a quiz.
- [ ] Add UC-03 integration tests for starting or reusing a quiz attempt.
- [ ] Add UC-03 integration tests for submitting a quiz attempt.
- [ ] Add UC-03 integration tests for retrieving quiz results.
- [ ] Verify quiz submission tests assert persisted attempt, answer, selected option, and result rows.
- [ ] Verify training progress tests assert persisted interaction rows.
- [ ] Add access-control tests for wrong trainee/campaign item/attempt scenarios.
- [ ] Add at least one not-found or forbidden database-backed path per major Demo 1 flow.
- [ ] Verify integration tests reset or isolate database state between tests.
- [ ] Update CI to run backend integration tests against a PostgreSQL test database.
- [ ] Ensure CI has separate, clearly named backend unit test and backend integration test steps/checks.
- [ ] Ensure CI makes it obvious whether a failure came from unit tests or integration tests.
- [ ] Ensure CI sets `TEST_DATABASE_URL` to a disposable test-only database.
- [ ] Ensure CI runs Prisma generation before integration tests.
- [ ] Ensure CI applies committed migrations before integration tests.
- [ ] Run backend unit tests.
- [ ] Run backend integration tests.
- [ ] Run backend typecheck.
- [ ] Update docs/scripts if another developer needs new instructions to run integration tests locally or in CI.

## Deadlines

Target completion: Before Demo 1 practice (Thursday)

## Acceptance Criteria

- [ ] `pnpm --filter @insightful-phish/backend test:integration` runs successfully against a real test database.
- [ ] Integration tests use `TEST_DATABASE_URL` or another clearly test-only database URL.
- [ ] Integration tests do not mock Prisma.
- [ ] Test database cleanup remains guarded against non-test databases.
- [ ] Committed Prisma migrations are usable by the integration test database.
- [ ] Auth register/login/me flows are covered against the real database.
- [ ] UC-01 simulated inbox/email flows are covered against the real database.
- [ ] UC-01 interaction/classification persistence is verified where implemented.
- [ ] UC-02 training document viewed/completed flows are covered against the real database.
- [ ] UC-03 quiz get/start/submit/results flows are covered against the real database.
- [ ] Quiz submission tests verify persisted attempt, answer, selected option, and result data where applicable.
- [ ] Training progress tests verify persisted viewed/completed interaction data.
- [ ] Access-control failures are tested with real relational data.
- [ ] Tests use realistic related data rather than only isolated single-table inserts.
- [ ] Integration tests reset or isolate database state between tests.
- [ ] `pnpm --filter @insightful-phish/backend test:unit` still passes.
- [ ] `pnpm --filter @insightful-phish/backend typecheck` still passes.
- [ ] CI provisions or connects to a PostgreSQL test database for backend integration tests.
- [ ] CI sets `TEST_DATABASE_URL` to a test-only database name accepted by the cleanup guard.
- [ ] CI runs `pnpm --filter @insightful-phish/backend prisma:generate` before integration tests.
- [ ] CI applies Prisma migrations before running integration tests.
- [ ] CI runs `pnpm --filter @insightful-phish/backend test:integration`.
- [ ] CI reports backend unit tests and backend integration tests as separate steps/checks.
- [ ] CI failure output makes it clear whether backend unit tests failed or backend integration tests failed.
- [ ] No frontend integration or UI work is included in this chore.

### PR DESCRIPTION
**Pull request by RudolphLamp**
## Summary

This PR introduces comprehensive, database-backed integration testing for all Demo 1 flows. We took the backend from relying solely on mocked unit tests to having a fully functional integration testing suite. Instead of assuming the code works in isolation, this proves that our Express routes, Prisma services, and PostgreSQL database interact correctly under real-world conditions.

**Here is the detailed breakdown of exactly what we built:**

* **1. CI Pipeline & Infrastructure Setup**
  * **CI Pipeline Split:** Updated the GitHub Actions workflow to separate backend unit tests from integration tests. This ensures that if a unit test fails, we know it is a logic error, but if an integration test fails, it is a database or migration issue.
  * **PostgreSQL Container:** Configured CI to spin up a dedicated PostgreSQL service, run `prisma:generate`, and apply migrations (`prisma migrate deploy`) before the integration tests execute.
  * **Safety Guards:** Hardened the database cleanup script. It now actively checks the connection string and refuses to truncate tables if the database name contains production keywords or lacks the word "test".
  * **Test Data Factories:** Built type-safe factory functions (`createTrainee`, `createCampaign`, `createQuiz`, etc.) to dynamically generate clean, isolated relational trees for every single test block instead of relying on rigid, shared seed data.

* **2. Authentication & Training Progress (UC-02)**
  * **Auth Flows:** Wrote tests hitting `/register`, `/login`, and `/me`. Verified that `User`, `TraineeProfile`, and `GeneralTraineeProfile` records actually write to the database and that passwords are functionally hashed using scrypt.
  * **Training Documents:** Tested the retrieval of training materials and hit the `/viewed` and `/completed` endpoints. Used the Prisma client to explicitly query the database and assert that `InteractionEvent` rows were successfully written with the correct campaign and target IDs.

* **3. Simulated Inbox Mechanics (UC-01)**
  * **Cheat Prevention:** Wrote tests ensuring that when a trainee requests their inbox or an email payload, sensitive data like correct classifications and red flags are successfully scrubbed from the response DTO.
  * **Interaction Tracking:** Proved that simulating a link click actively persists a `SIMULATED_EMAIL_LINK_CLICKED` event in the database.
  * **Classification Persistence:** Tested the submission of phishing/safe classifications. Asserted that the backend correctly calculates the score, persists an `EmailClassificationResponse`, links the selected red flags, and actively blocks duplicate submissions with a `409 Conflict`.

* **4. Quiz Generation & Grading (UC-03)**
  * **State Management:** Tested attempt initialization, verifying that the backend creates a `QuizAttempt` row marked as `IN_PROGRESS` and reuses that attempt ID rather than creating duplicates if called again.
  * **Grading and Submission:** Submitted quiz attempts and asserted that the backend correctly evaluates the answers. Checked the database to confirm the status shifts to `SUBMITTED`, and that `Answer` and `SelectedOption` rows are perfectly mapped to the database schema.
  * **Result Delivery:** Retrieved the finalized results, ensuring the system accurately calculates the percentage and pass/fail state based on the newly persisted relational data.

* **5. Access Control & Security**
  * **Cross-User Data Isolation:** Used the factories to spin up two completely distinct trainees. Proved that if Trainee A passes their Bearer token but tries to request Trainee B's simulated emails, quiz attempts, or training documents, the API strictly rejects it with a `403 Forbidden` or `404 Not Found`.
  * **Missing Entities:** Verified that passing invalid or non-existent UUIDs to the endpoints consistently returns a `404 Not Found` rather than crashing the server or throwing an unhandled 500 error.

## Linked Issue

Closes #132 

## Type of change

- [ ] Feature
- [ ] Bug Fix
- [ ] Documentation
- [x] Chore/Maintenance

## Testing

I successfully ran and verified the following locally to ensure no regressions:
* `pnpm --filter @insightful-phish/backend test:integration` (All 28 tests pass against the local test database).
* `pnpm --filter @insightful-phish/backend test:unit` (All 240 mocked unit tests still pass).
* `pnpm --filter @insightful-phish/backend typecheck` (No compiler errors).
* Simulated the CI environment locally to ensure migrations apply cleanly to a fresh test database before tests execute.

## Review Notes

* **Johan:** Please review the CI workflow changes in `.github/workflows/ci.yml` to ensure the PostgreSQL container setup aligns with our infrastructure standards and handles the migration steps properly.
* **Adriano:** Please take a look at the data factories in `apps/backend/tests/helpers/factories.ts` to confirm the Prisma relational mapping matches your expectations for the Demo 1 domain model.

## Checklist

- [x] I tested the change locally
- [x] No unrelated changes are included
- [x] Relevant tests were added or updated if applicable
- [x] Documentation was updated if needed
- [x] No secrets, credentials, or sensitive values were committed

### CODE REVIEW

File: apps/backend/tests/integration/uc01-inbox.integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  createEmailRedFlag,
+} from '../helpers/factories.js';
+import {
+  CampaignStatus,
+  CampaignComponentType,
+  InboxStatus,
+  EmailClassification,
+  EmailRedFlagType,
+  RedFlagSeverity,
+} from '../../src/generated/prisma/enums.js';
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/tests/integration/uc01-inbox.integration.test.ts:21]: OK BUDDY

File: apps/backend/tests/integration/access-control.integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
+import request from 'supertest';
+import { beforeEach, describe, expect, it } from 'vitest';
+import { createApp } from '../../src/app.js';
+import { prisma } from '../../src/lib/prisma.js';
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/tests/integration/access-control.integration.test.ts]: OK BUDDY

File: apps/backend/tests/integration/access-control.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+  createSimulation,
+  createSimulatedInbox,
+  createSimulatedEmail,
+  createQuiz,
+  createQuizQuestion,
+  createAnswerOption,
+  createCampaignItem,
+  createCampaignAssignment,
+  createTrainingDocument,
+} from '../helpers/factories.js';
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/tests/integration/access-control.integration.test.ts:15]: OK BUDDY

File: apps/backend/tests/integration/access-control.integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+import {
+  CampaignStatus,
+  CampaignComponentType,
+  QuizStatus,
+  InboxStatus,
+  EmailClassification,
+  TrainingDocumentStatus,
+  TrainingContentType,
+} from '../../src/generated/prisma/enums.js';
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/tests/integration/access-control.integration.test.ts:23]: OK BUDDY

**Comment by RudolphLamp** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=COS301-SE-2026_Cybersecurity-Awareness-Training-Platform&pullRequest=133) **Quality Gate failed**
> Failed conditions  [3.0% Duplication on New Code](https://sonarcloud.io/component_measures?id=COS301-SE-2026_Cybersecurity-Awareness-Training-Platform&pullRequest=133&metric=new_duplicated_lines_density&view=list) (required ≤ 3%)  [C Security Rating on New Code](https://sonarcloud.io/dashboard?id=COS301-SE-2026_Cybersecurity-Awareness-Training-Platform&pullRequest=133) (required ≥ A)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=COS301-SE-2026_Cybersecurity-Awareness-Training-Platform&pullRequest=133)
> 
>  Catch issues before they fail your Quality Gate with our IDE extension  [SonarQube for IDE](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=pull-request)

3e09f5dd35a0280039b88f0f7508f9479b97ffb408dc3969310be4d1f14fb5be_1

**Review by FJNel — CHANGES_REQUESTED** [review]
AAAA sorry I'm requesting one small but important change before approval.

The integration suite passes locally against the real `insightful_phish_test` database, and the CI/database setup looks safe: separate unit/integration jobs, PostgreSQL service, test-only DB URL, Prisma generate, and `migrate deploy` before integration tests. I also re-checked the auth rate-limit concern and no longer consider it blocking.

The remaining issue is in `apps/backend/tests/integration/access-control.integration.test.ts`: the test for “Trainee A attempts to complete a training document belonging to Trainee B” posts to `/trainee/campaign-items/:id/training-document/interactions`, but the backend only implements `/training-document/viewed` and `/training-document/completed`.

As implemented in your rbanch, the test can pass from a route-level 404 rather than proving access control. Please update that test to hit the real completion route, likely `/training-document/completed`, using Trainee A’s token against Trainee B’s campaign item.

**Comment by RudolphLamp** [comment]
Hi @FJNel,

I've updated the branch with the requested integration test additions and refactorings for #132. Could you please take a look and re-review?

**Review by FJNel — APPROVED** [review]
Everything seems to be good now and ready for approval! All tests also pass locally Thank you :D
